thrust/thrust/system/cuda/detail/dispatch.h

```diff
@@ -30,92 +30,145 @@
-  {                                                                                             \
-    auto THRUST_PP_CAT2(count1, _fixed) = static_cast<std::int32_t>(count1);                    \
-    auto THRUST_PP_CAT2(count2, _fixed) = static_cast<std::int32_t>(count2);                    \
-    status                              = call_32 arguments;                                    \
-  }                                                                                             \
-  else                                                                                          \
-  {                                                                                             \
-    auto THRUST_PP_CAT2(count1, _fixed) = static_cast<std::int64_t>(count1);                    \
-    auto THRUST_PP_CAT2(count2, _fixed) = static_cast<std::int64_t>(count2);                    \
-    status                              = call_64 arguments;                                    \
-  }
+#if defined(THRUST_FORCE_64BIT_OFFSET_TYPE)
+//! @brief Always dispatches to 64 bit offset version of an algorithm
+#  define THRUST_INDEX_TYPE_DISPATCH(status, call, count, arguments) \
+    _THRUST_INDEX_TYPE_DISPATCH(std::int64_t, status, call, count, arguments)
+
+//! Like \ref THRUST_INDEX_TYPE_DISPATCH but with two counts
+#  define THRUST_DOUBLE_INDEX_TYPE_DISPATCH(status, call, count1, count2, arguments) \
+    _THRUST_INDEX_TYPE_DISPATCH2(std::int64_t, status, call, count1, count2, arguments)
+
+//! Like \ref THRUST_INDEX_TYPE_DISPATCH but with two different call implementations
+#  define THRUST_INDEX_TYPE_DISPATCH2(status, call_32, call_64, count, arguments) \
+    _THRUST_INDEX_TYPE_DISPATCH(std::int64_t, status, call_64, count, arguments)
+
+//! Like \ref THRUST_INDEX_TYPE_DISPATCH2 but uses two counts.
+#  define THRUST_DOUBLE_INDEX_TYPE_DISPATCH2(status, call_32, call_64, count1, count2, arguments) \
+    _THRUST_INDEX_TYPE_DISPATCH2(std::int64_t, status, call_64, count1, count2, arguments)
+
+//! Like \ref THRUST_INDEX_TYPE_DISPATCH2 but always dispatching to uint64_t. `count` must not be negative.
+#  define THRUST_UNSIGNED_INDEX_TYPE_DISPATCH2(status, call_32, call_64, count, arguments) \
+    _THRUST_INDEX_TYPE_DISPATCH(std::uint64_t, status, call_64, count, arguments)
+
+#elif defined(THRUST_FORCE_32BIT_OFFSET_TYPE)
+
+//! @brief Ensures that the size of the input does not overflow the offset type
+#  define _THRUST_INDEX_TYPE_DISPATCH_GUARD_OVERFLOW(index_type, count)                      \
+    if (static_cast<std::uint64_t>(count)                                                    \
+        > static_cast<std::uint64_t>(thrust::detail::integer_traits<index_type>::const_max)) \
+    {                                                                                        \
+      throw ::std::runtime_error("Offset type overflow");                                    \
```

Comment:
Let's make the error more descriptive. 
```suggestion
      throw ::std::runtime_error(                                                            \
        "Input size exceeds the maximum allowable value for " #index_type                    \
        " (" + std::to_string(thrust::detail::integer_traits<index_type>::const_max) + "). " \
        #index_type " was used because the macro THRUST_FORCE_32BIT_OFFSET_TYPE was defined. "\
        "To handle larger input sizes, either remove this macro to dynamically dispatch "    \
        "between 32-bit and 64-bit index types, or define THRUST_FORCE_64BIT_OFFSET_TYPE.")
```